xarray/backends/rasterio_.py

```diff
@@ -0,0 +1,166 @@
+    """Open a file with rasterio (experimental).
+
+    This should work with any file that rasterio can open (most often:
+    geoTIFF). The x and y coordinates are generated automatically from the
+    file's geoinformation.
+
+    Parameters
+    ----------
+    filename : str
+        Path to the file to open.
+
+    Returns
+    -------
+    data : DataArray
+        The newly created DataArray.
+    chunks : int, tuple or dict, optional
+        Chunk sizes along each dimension, e.g., ``5``, ``(5, 5)`` or
+        ``{'x': 5, 'y': 5}``. If chunks is provided, it used to load the new
+        DataArray into a dask array.
+    cache : bool, optional
+        If True, cache data loaded from the underlying datastore in memory as
+        NumPy arrays when accessed to avoid reading from the underlying data-
+        store multiple times. Defaults to True unless you specify the `chunks`
+        argument to use dask, in which case it defaults to False.
+    lock : False, True or threading.Lock, optional
+        If chunks is provided, this argument is passed on to
+        :py:func:`dask.array.from_array`. By default, a global lock is
+        used to avoid issues with concurrent access to the same file when using
+        dask's multithreaded backend.
+    """
+
+    import rasterio
+    riods = rasterio.open(filename, mode='r')
+
+    if cache is None:
+        cache = chunks is None
+
+    coords = OrderedDict()
+
+    # Get bands
+    if riods.count < 1:
+        raise ValueError('Unknown dims')
+    coords['band'] = np.asarray(riods.indexes)
+
+    # Get geo coords
+    nx, ny = riods.width, riods.height
+    dx, dy = riods.res[0], -riods.res[1]
+    x0 = riods.bounds.right if dx < 0 else riods.bounds.left
+    y0 = riods.bounds.top if dy < 0 else riods.bounds.bottom
+    coords['y'] = np.linspace(start=y0, num=ny, stop=(y0 + (ny - 1) * dy))
+    coords['x'] = np.linspace(start=x0, num=nx, stop=(x0 + (nx - 1) * dx))
+
+    # Attributes
+    attrs = {}
+    if hasattr(riods, 'crs'):
+        # CRS is a dict-like object specific to rasterio
+        # We convert it back to a PROJ4 string using rasterio itself
+        attrs['crs'] = riods.crs.to_string()
+    # Maybe we'd like to parse other attributes here (for later)
+
+    data = indexing.LazilyIndexedArray(RasterioArrayWrapper(riods))
+
+    # this lets you write arrays loaded with rasterio
+    data = indexing.CopyOnWriteArray(data)
+    if cache and (chunks is None):
+        data = indexing.MemoryCachedArray(data)
+
+    result = DataArray(data=data, dims=('band', 'y', 'x'),
+                       coords=coords, attrs=attrs)
+
+    if chunks is not None:
+        from dask.base import tokenize
+        # augment the token with the file modification time
+        mtime = os.path.getmtime(filename)
+        token = tokenize(filename, mtime, chunks)
+        name_prefix = 'open_rasterio-%s' % token
+        if lock is None:
+            lock = GLOBAL_LOCK
```

Comment:
Actually, can you make a *separate* global lock for rasterio? We can safely use GDAL and HDF5 at the same time.